### PR TITLE
Add runBenchmarkWith, general benchmark runner

### LINKED ISSUE #254: Add a variant of runBenchmark which also takes iteration range.
Currently there is:

```haskell
runBenchmark :: Benchmarkable -> Double -> IO (Vector Measured, Double)
```

> Run a single benchmark, and return measurements collected while executing it, along with the amount of time the measurement process took.

> Double argument is: Lower bound on how long the benchmarking process should take.  In practice, this time limit may be exceeded in order to generate enough data to perform meaningful statistical analyses.

I'd like to have a variant where we can specify the optional lower and upper bounds for iteration count (run at least 10 times, no need to run over 100 times).

Lower bound would be great to ensure more statistical significance. Upper bound would allow to avoid running individual benchmark for tool long.

I think that then we'd need both lower and upper bound for time as well (current is actually more like soft upper bound, but as there are no end criteria it's also a lower bound).

I think it would best to put these four options into a record, to avoid double-blindness.
In case when no bounds is specified, we could either fail or use some default time bound.

### PR DESCRIPTION
**Pull request by phadej**
Refactor runBenchmark to use it.
Resolves #254

---

I opted out to generating some data-type with configuration etc. The logic is just easier to write as code directly.

### CODE REVIEW

**Comment by phadej** [comment]
I'll add semigroups compat if this is otherwise ok, or when I make tweaks.

File: criterion-measurement/src/Criterion/Measurement.hs [hunk 1]
```diff
@@ -289,41 +291,69 @@ runBenchmark :: Benchmarkable
              -- exceeded in order to generate enough data to perform
              -- meaningful statistical analyses.
              -> IO (V.Vector Measured, Double)
-runBenchmark bm timeLimit = do
+runBenchmark bm timeLimit = runBenchmarkWith endAt 0 bm where
+  endAt :: Int -> Int64 -> Double -> NonEmpty Measured -> Double -> Maybe (Int64, Double)
```
Comment (RyanGlScott) [inline criterion-measurement/src/Criterion/Measurement.hs:295]: Since this is the default behavior for `runBenchmarkWith`'s termination check, I'd favor factoring this out into a top-level definition so that a user can use it if they wish to tweak it slightly. Since this definition closes over `timeLimit`, it might make things more clear to turn this into a newtype:

```hs
newtype TerminationCheck s = TerminationCheck
  { getTerminationCheck :: Int -> Int64 -> Double -> NonEmpty Measured -> s -> Maybe (Int64, s) }

defaultTerminationCheck :: Double -> TerminationCheck s
defaultTerminationCheck timeLimit = TerminationCheck $ \count iters delta ms prev -> do ...
```

Then you can mention in the Haddocks for `runBenchmark` that `runBenchmark bm timeLimit = runBenchmarkWith (defaultTerminationCheck timeLimit) 0 bm`.

File: criterion-measurement/src/Criterion/Measurement.hs [hunk 2]
```diff
@@ -289,41 +291,69 @@ runBenchmark :: Benchmarkable
...
+-- * current run count (starting from zero)
+-- * current iteration count (starting from one)
+-- * amount of time the measurement process took so far.
+-- * results so far (in reverse order, latest first in the list).
+-- * user defined state
+--
+-- and should return 'Nothing' for run to terminate
+-- or @'Just' (nextIters, nextState)@ for run to continue with new iteration count and state.
+--
+runBenchmarkWith :: forall s. (Int -> Int64 -> Double -> NonEmpty Measured -> s -> Maybe (Int64, s))
```
Comment (RyanGlScott) [inline criterion-measurement/src/Criterion/Measurement.hs:334]: Speaking of which, what _is_ the best way to tweak the default behavior? I imagine that a common use case would be to use `runBenchmarkWith` to impose a lower maximum on the number of iterations being run, as per #210/#218. Is this straightforward to accomplish with this design?

File: criterion-measurement/src/Criterion/Measurement.hs [hunk 3]
```diff
@@ -289,41 +291,69 @@ runBenchmark :: Benchmarkable
...
+-- The first argument is termination check, it takes as arguments:
+--
+-- * current run count (starting from zero)
+-- * current iteration count (starting from one)
+-- * amount of time the measurement process took so far.
+-- * results so far (in reverse order, latest first in the list).
+-- * user defined state
+--
+-- and should return 'Nothing' for run to terminate
+-- or @'Just' (nextIters, nextState)@ for run to continue with new iteration count and state.
```
Comment (RyanGlScott) [inline criterion-measurement/src/Criterion/Measurement.hs:332]: This needs a loud disclaimer that you should be careful when picking a termination check, as its implementation could affect the statistical quality of the data.

File: criterion-measurement/src/Criterion/Measurement.hs [hunk 4]
```diff
@@ -289,41 +291,69 @@ runBenchmark :: Benchmarkable
...
+
+  -- Multiply by 1.05, but always increase
+  nextIters :: Int64 -> Int64
+  nextIters n = max (n + 1) (div (105 * n) 100)
+
+-- | Run benchmark with provided dynamic termination condition.
+--
+-- The first argument is termination check, it takes as arguments:
+--
+-- * current run count (starting from zero)
```
Comment (RyanGlScott) [inline criterion-measurement/src/Criterion/Measurement.hs:325]: If `TerminationCheck` were factored out into a newtype, these Haddocks could be moved to the newtype instead.

File: criterion-measurement/src/Criterion/Measurement.hs (hunk 1 above)
Comment (phadej) [reply criterion-measurement/src/Criterion/Measurement.hs:295]: I'm not so sure about. That function needs to be factored so its pieces could be  usable, but I don't know how we could factor it. There are at least two parts: updating the state and deciding termination, but that consist of three checks already.

File: criterion-measurement/src/Criterion/Measurement.hs (hunk 2 above)
Comment (phadej) [reply criterion-measurement/src/Criterion/Measurement.hs:334]: I don't see an other way then copying the source and tweaking it.

File: criterion-measurement/src/Criterion/Measurement.hs (hunk 3 above)
Comment (phadej) [reply criterion-measurement/src/Criterion/Measurement.hs:332]: [`analyseSample`](https://hackage.haskell.org/package/criterion-1.5.12.0/docs/Criterion-Analysis.html#v:analyseSample) doesn't have any pre condiotions specified. Looking at the code, it apperently needs at least 4 samples, then the statistical quality will be estimated (in the confidence interval).


I.e. I don't know how to put it "Termination check condition should be picked so the data generated is suitable for whatever further analysis is performed on it". Isn't that obvious?

**Comment by phadej** [comment]
I changed the type of function. It fits better the usage where some analysis is done already on each step (in ~bayesian style), so you won't end doing it twice on the last sample (and without access to the state).

File: criterion-measurement/src/Criterion/Measurement.hs (hunk 1 above)
Comment (RyanGlScott) [reply criterion-measurement/src/Criterion/Measurement.hs]: I think I explained myself poorly in https://github.com/haskell/criterion/pull/255#discussion_r783494620, so let me try again. I agree that because termination checks will need to perform arbitrary computation, there's not a good way in general to configure everything that a termination check might want to do. That being said, I still think there is value in exposing the default termination check to users. This is because in #218, a user wishes to do what `runBenchmark` does plus an additional constraint on the number of iterations performed. In terms of code, this could look something like this:

```hs
longLivedComputationCheck :: TerminationCheck r s
longLivedComputationCheck = TerminationCheck $ \... -> do
  getTerminationCheck defaultTerminationCheck ...
  if <time exceeds certain threshold>
    then Left ...
    else Right ...
```

For some use cases, there might not be a better alternative than having to copy-paste the source code of `defaultTerminationCheck` and tweaking it, but I don't want that to always be the case.

File: criterion-measurement/src/Criterion/Measurement.hs (hunk 3 above)
Comment (RyanGlScott) [reply criterion-measurement/src/Criterion/Measurement.hs:335]: To be clear, I'm not looking for a super-detailed analysis of _how_ to make the data statistically significant. I'm just looking for a generic warning saying that you should exercise caution when using this function, as it has the potential to distort the meaningfulness of the data if implemented haphazardly.

**Comment by phadej** [comment]
I don't need this functionality anymore.
